src/storage/src/hummock/file_cache/cache.rs

```diff
@@ -0,0 +1,392 @@
+
+                for ((key, value), slot) in batch.into_iter().zip_eq(slots.into_iter()) {
+                    let hash = self.hash_builder.hash_one(&key);
+                    self.indices.insert(key, hash, value.len(), slot);
+                }
+
+                self.buffer.rotate();
+            }
+
+            for hook in &self.hooks {
+                hook.post_flush().await?;
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct FileCache<K, S = RandomState>
+where
+    K: CacheKey,
+    S: HashBuilder,
+{
+    hash_builder: S,
+
+    _filters: Vec<Arc<dyn Filter>>,
+
+    indices: Arc<LruCache<K, SlotId>>,
+
+    store: StoreRef<K>,
+
+    buffer: TwoLevelBuffer<K>,
+    buffer_flusher_notifier: Arc<Notify>,
+}
+
+impl<K> FileCache<K, RandomState>
+where
+    K: CacheKey,
+{
+    pub async fn open(options: FileCacheOptions) -> Result<Self> {
+        let hash_builder = RandomState::new();
+        Self::open_with_hasher(options, hash_builder).await
+    }
+}
+
+impl<K, S> FileCache<K, S>
+where
+    K: CacheKey,
+    S: HashBuilder,
+{
+    pub async fn open_with_hasher(options: FileCacheOptions, hash_builder: S) -> Result<Self> {
+        let buffer_capacity = options.total_buffer_capacity / 2;
+
+        let store = Store::open(StoreOptions {
+            dir: options.dir,
+            capacity: options.capacity,
+            buffer_capacity,
+            cache_file_fallocate_unit: options.cache_file_fallocate_unit,
+        })
+        .await?;
+        let store = Arc::new(store);
+
+        // TODO: Restore indices.
+        let indices =
+            LruCache::with_event_listeners(LRU_SHARD_BITS, options.capacity, vec![store.clone()]);
+        store.restore(&indices).await?;
+        let indices = Arc::new(indices);
+
+        let buffer = TwoLevelBuffer::new(buffer_capacity);
+        let buffer_flusher_notifier = Arc::new(Notify::new());
+
+        let buffer_flusher = BufferFlusher {
+            buffer: buffer.clone(),
+            store: store.clone(),
+            indices: indices.clone(),
+            notifier: buffer_flusher_notifier.clone(),
+
+            hash_builder: hash_builder.clone(),
+
+            hooks: options.flush_buffer_hooks,
+        };
+        // TODO(MrCroxx): Graceful shutdown.
+        let _handle = tokio::task::spawn(async move {
+            if let Err(e) = buffer_flusher.run().await {
+                tracing::error!("error raised within file cache buffer flusher: {}", e);
+            }
+        });
+
+        Ok(Self {
+            hash_builder,
+
+            _filters: options.filters,
+
+            indices,
+
+            store,
+
+            buffer,
+            buffer_flusher_notifier,
+        })
+    }
+
+    pub fn insert(&self, key: K, value: Vec<u8>) -> Result<()> {
+        let hash = self.hash_builder.hash_one(&key);
+        self.buffer.insert(hash, key, value.len(), value);
+
+        self.buffer_flusher_notifier.notify_one();
```

Comment:
The `Notify` struct can only store at most one signal. So, when the flusher is busy flushing the last buffer, it will be triggered only once if there are multiple inserts in the next loop.

I think the model is okay because we don't have many other modules use EBS. So, trying to fully utilizing EBS is by design. A free EBS will not do any benefits.